File: leetcode2/3hard/최원준/Q1402.java
package Leetcode;

/*
1. 아이디어 :
curr : 현재까지의 만족도
total에 누적된 curr의 합을 저장하고, ㅁns를 갱신

2. 시간복잡도 :
O( n )

3. 자료구조/알고리즘 :
- / -
 */

import java.util.Arrays;

public class Q1402 {
    class Solution {
        public int maxSatisfaction(int[] satisfaction) {
            Arrays.sort(satisfaction);
            int n = satisfaction.length;
            int curr = 0, total = 0, ans = 0;
            for (int i=n-1; i>-1; i--) {
                curr += satisfaction[i];
                total += curr;
                ans = Math.max(ans, total);
            }
            return ans;
        }
    }
}

File: leetcode2/3hard/강병호/Q1402.py
class Leetcode.이재훈.이재훈.Solution:
    def maxSatisfaction(self, satisfaction: list[int]) -> int:
        # n 개 dishes
        # like-time coefficient : time[i] * satisfaction[i]
        # dish 삭제하여 최대

        satisfaction.sort()

        start = 0
        result = 0
        while start != len(satisfaction):
            sum_satis = 0
            time=1
            for i in range(start, len(satisfaction)):
                sum_satis += time * satisfaction[i]
                time+=1
            

            print(f"sum_satis : {sum_satis}")
            result = max(result, sum_satis)
            print(f"result : {sum_satis}")

            start += 1
        
        return result


if __name__ == "__main__":
    sol = Leetcode.이재훈.이재훈.Solution()
    print(sol.maxSatisfaction([-1, -8, 0, 5, -9]))
    print(sol.maxSatisfaction([4,3,2]))
    print(sol.maxSatisfaction([-1,-4,-5]))
